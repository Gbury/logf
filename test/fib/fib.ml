
(* This file is free software, part of logf. See file "LICENSE" for more information *)

open Logf

let rec fib n =
  withf "Starting fib(%d)" n
    ~key:(string_of_int n) ~data:[| "n", Int n; |] ~f:(fun () ->
        match n with
        | 0 -> 0
        | 1 -> 1
        | _ -> fib (n - 1) + fib (n - 2)
      )

let main n =
  logf "Initialization...";
  let res = fib n in
  logf "Finished !"
    ~data:[|"res", Int res|];
  ()

let () =
  let n = ref 0 in
  Arg.parse [
  ] (fun anon_arg ->
      n := int_of_string anon_arg)
    {|fib <n> : example of nested logging using the fibonacci function|};
  main !n

