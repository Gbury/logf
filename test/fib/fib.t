
Run the fib executable to generate some logs
  $ ./fib.exe 5 &> fib.logf

Check that the logs are parseable
  $ logf -- fib.logf
  [1]<>:
  [2]<5>:
    [3]<3>:
      [4]<1>:
      [5]<2>:
        [6]<0>:
        [7]<1>:
    [8]<4>:
      [9]<2>:
        [10]<0>:
        [11]<1>:
      [12]<3>:
        [13]<1>:
        [14]<2>:
          [15]<0>:
          [16]<1>:
  [17]<>:

Try out some queries on nodes when we know the id
  $ logf query fib.logf 1
  <1> :
    n: 5

  $ logf query fib.logf 8
  <8> 5:4:
    n: 4

  $ logf query fib.logf 12
  <12> 5:4:3:
    n: 3

Try and find some paths based on some regexps
  $ logf find fib.logf "**:3:**"
  No paths match the given regular expression
