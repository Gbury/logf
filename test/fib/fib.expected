[00001] Initialization... 
[^0002] 5  Starting fib(5)
   n  :  5 
[^0003] 3  Starting fib(3)
   n  :  3 
[^0004] 1  Starting fib(1)
   n  :  1 
[$0004]  
[^0005] 2  Starting fib(2)
   n  :  2 
[^0006] 0  Starting fib(0)
   n  :  0 
[$0006]  
[^0007] 1  Starting fib(1)
   n  :  1 
[$0007]  
[$0005]  
[$0003]  
[^0008] 4  Starting fib(4)
   n  :  4 
[^0009] 2  Starting fib(2)
   n  :  2 
[^0010] 0  Starting fib(0)
   n  :  0 
[$0010]  
[^0011] 1  Starting fib(1)
   n  :  1 
[$0011]  
[$0009]  
[^0012] 3  Starting fib(3)
   n  :  3 
[^0013] 1  Starting fib(1)
   n  :  1 
[$0013]  
[^0014] 2  Starting fib(2)
   n  :  2 
[^0015] 0  Starting fib(0)
   n  :  0 
[$0015]  
[^0016] 1  Starting fib(1)
   n  :  1 
[$0016]  
[$0014]  
[$0012]  
[$0008]  
[$0002]  
[00017] Finished !
   res  :  5 
